### PR TITLE
Implementations IEquatable<T> for custom structs.

### PR DESCRIPTION
**Pull request by sq735**
Motivation:

Best practices recommend implementing an "IEquatable<T>" interface for custom structs. This improves the performance of comparison operations and eliminates boxing.

### CODE REVIEW

**Comment by nblumhardt** [comment]
Thanks for the PR! Is there anywhere currently in the code path that we'd expect this to speed up?

Also, especially if so, should we first consider making these types into C#10 record structs?

(I haven't dug into this in quite some time, so please excuse some possibly obvious questions :-))

**Comment by nblumhardt** [comment]
Hi! Thanks for your reply. Giving this some thought, it would probably be a bug if we were using struct equality implicitly anywhere in this codebase, so adding support for it may be confusing (we don't generally want to use it, so the impls there would be misleading). Let me know if this matches your impressions.

Best regards,
Nick

**Comment by sq735** [comment]
I don't think an explicit implementation would be misleading. But if you are of a different opinion, then the `record struct` are your choice. Since they do not require an explicit implementation of equivalence and eliminate a number of problems inherent in classical structures.

**Comment by nblumhardt** [comment]
Thanks, Pavel - giving it some more thought 👍

**Comment by sq735** [comment]
Nicholas, you can see some performance comparison for structs and record struct [here](https://nietras.com/2021/06/14/csharp-10-record-struct/).

File: src/Superpower/Model/Position.cs [hunk 1]
```diff
@@ -88,5 +88,60 @@ public override string ToString()
...
+        public override int GetHashCode()
+        {
+            unchecked
+            {
+                var hashCode = Absolute;
+                hashCode = (hashCode * 397) ^ Line;
+                hashCode = (hashCode * 397) ^ Column;
+                return hashCode;
+            }
+        }
```
Comment (0xF6) [inline src/Superpower/Model/Position.cs:125]: Please use [HashCode.Combine](https://learn.microsoft.com/dotnet/api/system.hashcode?view=net-8.0)
```suggestion
        public override int GetHashCode() => HashCode.Combine(Absolute, Line, Column);
```

**Comment by nblumhardt** [comment]
Thanks again for sending this. I've run back over everything and decided not to merge this one in order to keep the codebase lean. Although there's some hypothetical benefit, these structs are not intended to be used as comparable values and in current usage the added methods won't be used.

In the case of `Position`, which is a reasonable candidate for use as a comparable value, the different types of comparisons are nuanced, and in the cases we care about, only `Absolute` is compared (since the counted line/column info has no bearing on the case of identifying a span within a string).

If comparison using one of these types does show up in a scenario somewhere we can definitely revisit. Thanks again @sq735!

**Pull request closed without merging**